junit-platform-launcher/src/main/java/org/junit/platform/launcher/listeners/MutableTestExecutionSummary.java

```diff
@@ -213,13 +216,46 @@ private void printSource(PrintWriter writer, TestIdentifier testIdentifier) {
 	}
 
 	private void printStackTrace(PrintWriter writer, Throwable throwable, int max) {
-		StackTraceElement[] elements = throwable.getStackTrace();
-		for (int i = 0; i < (elements.length > max ? max : elements.length); i++) {
-			writer.printf("%s   %s%n", DOUBLE_TAB, elements[i]);
+		if (throwable.getCause() != null
+				|| (throwable.getSuppressed() != null && throwable.getSuppressed().length > 0)) {
+			max = max / 2;
+		}
+		printStackTrace(writer, new StackTraceElement[] {}, throwable, "", DOUBLE_TAB + " ", max);
+		writer.flush();
+	}
+
+	private void printStackTrace(PrintWriter writer, StackTraceElement[] parentTrace, Throwable throwable,
+			String caption, String indentation, int max) {
+		StackTraceElement[] trace = throwable.getStackTrace();
+		if (parentTrace != null && parentTrace.length > 0) {
+			writer.printf("%s%s%s%n", indentation, caption, throwable);
+		}
+		int duplicates = numberOfCommonFrames(trace, parentTrace);
+		int numLines = (trace.length - duplicates > max) ? max : trace.length - duplicates;
```

Comment:
Nit: any good name for the difference `trace.length - duplicates`? Could be assigned to local variable used twice here.